### PR TITLE
query: Added a new flag --required-matcher

### PR DESCRIPTION
**Pull request by Harshitha1234**
Signed-off-by: Harshitha1234 <harshithachowdary.t17@iiits.in>

* [ ] I added CHANGELOG entry for this change.
* [ ] Change is not relevant to the end user.

## Changes
#2405 Task 2: Added a new flag `--required-matcher` in the querier

## Verification

Tested it locally on dev environment

### CODE REVIEW

File: pkg/store/proxy.go [hunk 1]
```diff
@@ -198,9 +201,37 @@ func (s ctxRespSender) send(r *storepb.SeriesResponse) {
 	s.ch <- r
 }
 
+func matchesRequiredMatchers(ms []storepb.LabelMatcher, requiredMatcher labels.Labels) (bool, error) {
```
Comment (yeya24) [inline pkg/store/proxy.go:204]: This function is almost the same as the function `matchesExternalLabels` below.  What about using that function directly? Maybe we can rename `matchesExternalLabels` to sth else. WDYT?

Comment (Harshitha1234) [reply pkg/store/proxy.go:204]: Yeah we can use `matchesExternalLabels` with the following logic
```go
ok, newmatchers, err := matchesExternalLabels(r.Matchers, s.requiredMatcher)
if err != nil {
	return status.Error(codes.InvalidArgument, err.Error())
}
if !ok {
	return nil
}
if len(newmatchers) > len(r.Matchers)-len(s.requiredMatcher) {
	return nil
}
```
WDYT?

File: cmd/thanos/query.go [hunk 2]
```diff
@@ -74,6 +74,9 @@ func registerQuery(m map[string]setupFunc, app *kingpin.Application) {
 	selectorLabels := cmd.Flag("selector-label", "Query selector labels that will be exposed in info endpoint (repeated).").
 		PlaceHolder("<name>=\"<value>\"").Strings()
 
+	requiredMatcher := cmd.Flag("required-matcher", "If required matchers are specified in a querier then the querier processes a query only if these labels are present in the query request.").
```
Comment (bwplotka) [inline cmd/thanos/query.go:77]:
```suggestion
	requiredMatcher := cmd.Flag("required-matcher", "Matchers that will be required in order to processes any request to `Series` (or `Select`  via query HTTP). If request matches will not have matchers required, query will return error. (repeated).").
```

File: pkg/store/proxy.go (hunk 1 above)
Comment (bwplotka) [reply pkg/store/proxy.go:204]: Hm, I don't think this is the same. This is totally agnostic to external matchers at current state.

**Comment by dgl** [comment]
@bwplotka yes, I think this is useful. It may be worth marking the flag as "experimental" as the more useful per store config will look a bit different.

File: cmd/thanos/query.go [hunk 3]
```diff
@@ -74,6 +74,9 @@ func registerQuery(m map[string]setupFunc, app *kingpin.Application) {
 	selectorLabels := cmd.Flag("selector-label", "Query selector labels that will be exposed in info endpoint (repeated).").
 		PlaceHolder("<name>=\"<value>\"").Strings()
 
+	requiredMatcher := cmd.Flag("required-matcher", "Matchers that will be required in order to processes any request to `Series` (or `Select`  via query HTTP). If request matches will not have matchers required, query will return error. (repeated).").
```
Comment (bwplotka) [inline cmd/thanos/query.go:77]: `.Hidden()`

```suggestion
	requiredMatcher := cmd.Flag("required-matcher", "Experimental: Matchers that will be required in order to processes any request to `Series` (or `Select`  via query HTTP). If request matches will not have matchers required, query will return error. (repeated).").
```

+ Let's make it matcher (:

**Comment by Harshitha1234** [comment]
Maybe we could implement this at the storeAPI level in the component where we are using this feature (For example., Geras) as it could be much better than the current feature at per Series level  and maybe we could have the required-matcher feature at per Store level which would really be quite helpful as @dgl suggested. WDYT?

**Comment by bwplotka** [comment]
Yea, we re-review this work and TBH as @Harshitha1234 said, this small feature is something you can totally do now in Geras, right @dgl ? TBH it looks like this feature would be (until we have per client configuration) only for your use case :hugs: (:

**Pull request closed without merging**